### DISCUSSION TITLE
Creating an Operator to manage function lifecycle (building, deploying and possibly invoke)

### DISCUSSION THREAD

**Discussion by luis5tb** [knative/func#1624] [Prototype Ideas]
I would like to get some inputs for the usefulness of creating an operator to manage functions lifecycle, i.e., to pretty much do what the "kn func XXX" does today. The main use case we have for that is for a multicluster setup, managed with Open Cluster Management (OCM), from where we want to be able to build, deploy and run different functions on different clusters. The idea is to trigger that through OCM ManifestWork, that then gets applied on the related managed cluster. 

Thanks to having this new operator running on the managed clusters, when a CR object of this new CRD type is created, the operator does the needed actions to build and/or deploy (and perhaps and/or invoke) the function described in it.

If this is a useful first starting point, I would like to know the feasibility of this, and get some help defining the CRD fields so that they are valid for other use cases that you see fit

**Comment by aslom** [comment]
Maybe a naive question: for that operator to work where would be the source code delcared in CR? Would it be in github repository and then cloned inside operator (assuming gitub repo is public) then `func` is called on that directory? 

Or the operator maybe creating a build pipeline and triggering it to build from source code and deploy into namespace where is CR?

**Reply by luis5tb** [reply]
Not sure I fully understand. Idea is that the CR will have a pointer to the code (public github repo), or the container image if already build.

Then, for the second point, yeah, I'm not sure if it is better to have this operator having `func` inside or mimic the pipeline logic to actually build from source.

Regarding the deploy, we split the process in two steps:
- building: creating the image for the function in the local/remote registry
- run: deploy the application on the cluster
- invoke: calling that function/service

So, we can have a flag to deploy or not (aka run) the function after it was build

**Comment by lkingland** [comment]
It looks like I may have duplicated your issue while creating discussions for each of the larger initiatives on our Functions Roadmap!

See #2088 

> I would like to know the feasibility of this, 

This is quite feasable, especially now that we have on-cluster builds.

> and get some help defining the CRD fields

Since the entirety of a Function's state is tracked as code, and expected to be in a repository, the only field required is the URL at which the source code can be found.

> I'm not sure if it is better to have this operator having func inside or mimic the pipeline logic

It should probably have `func` or import the client library and use it directly

> So, we can have a flag to deploy or not (aka run) the function after it was build

`run` and `invoke` are both development "helper" commands, and `build` is done automatically on `deploy`.  Therefore the only command the operator would need to execute would be `deploy`.
